sycl/doc/extensions/experimental/sycl_ext_oneapi_virtual_mem.asciidoc

```diff
@@ -0,0 +1,347 @@
+
+size_t get_recommended_mem_granularity(const device &syclDevice, const context &syclContext);
+size_t get_recommended_mem_granularity(const queue &syclQueue);
+size_t get_recommended_mem_granularity(const physical_mem &syclPhysicalMem);
+
+} // namespace sycl::ext::oneapi::experimental
+```
+
+[frame="topbot",options="header,footer"]
+|=====================
+|Function |Description
+
+|`size_t get_minimum_mem_granularity(const device &syclDevice, const context &syclContext)` |
+Returns the minimum granularity of physical and virtual memory allocations on
+`syclDevice` in the `syclContext`.
+
+If `syclDevice` does not have `aspect::ext_oneapi_virtual_mem` the call throws
+an exception with `errc::feature_not_supported`.
+
+|`size_t get_minimum_mem_granularity(const queue &syclQueue)` |
+Same as `get_minimum_mem_granularity(syclQueue.get_device(), syclQueue.get_context())`.
+
+|`size_t get_minimum_mem_granularity(const physical_mem &syclPhysicalMem)` |
+Same as `get_minimum_mem_granularity(syclPhysicalMem.get_device(), syclPhysicalMem.get_context())`.
+
+|`size_t get_recommended_mem_granularity(const device &syclDevice, const context &syclContext)` |
+Returns the recommended granularity of physical and virtual memory allocations
+on `syclDevice` in the `syclContext`.
+
+If `syclDevice` does not have `aspect::ext_oneapi_virtual_mem` the call throws
+an exception with `errc::feature_not_supported`.
+
+|`size_t get_recommended_mem_granularity(const queue &syclQueue)` |
+Same as `get_recommended_mem_granularity(syclQueue.get_device(), syclQueue.get_context())`.
+
+|`size_t get_recommended_mem_granularity(const physical_mem &syclPhysicalMem)` |
+Same as `get_recommended_mem_granularity(syclPhysicalMem.get_device(), syclPhysicalMem.get_context())`.
+
+|=====================
+
+=== Reserving virtual address ranges
+
+Virtual address ranges are represented by a pointer and a number of bytes
+reserved for it. The pointer must be aligned in accordance with the minimum
+granularity, as queried through `get_minimum_mem_granularity`, and likewise the
+number of bytes must be a multiple of this granularity. It is the responsibility
+of the user to manage the constituents of any virtual address range they
+reserve.
+
+The interfaces for reserving, freeing, and manipulating the access mode of a
+virtual address range are defined as:
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+
+uintptr_t reserve_virtual_mem(uintptr_t start, size_t numBytes, const context &syclContext);
+uintptr_t reserve_virtual_mem(size_t numBytes, const context &syclContext);
+
+void free_virtual_mem(uintptr_t ptr, size_t numBytes, const context &syclContext);
+
+} // namespace sycl::ext::oneapi::experimental
+```
+
+[frame="topbot",options="header,footer"]
+|=====================
+|Function |Description
+
+|`uintptr_t reserve_virtual_mem(uintptr_t start, size_t numBytes, const context &syclContext)` |
+Reserves a virtual memory range in `syclContext` with `numBytes` bytes.
```

Comment:
The [CUDA interface](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__VA.html#group__CUDA__VA_1ge489256c107df2a07ddf96d80c86cd9b) has the restriction
> The size and address parameters must be a multiple of the host page size and the alignment must be a power of two or zero for default alignment.

while the [Level Zero interface](https://spec.oneapi.io/level-zero/latest/core/api.html#zevirtualmemreserve) mentions the page size (used here as the granularity) as
> The starting address and size must be page aligned. See [zeVirtualMemQueryPageSize](https://spec.oneapi.io/level-zero/latest/core/api.html#ze__api_8h_1afe411cceb631a3a176463ee05fda3dd7).

Neither takes a device, despite the granularity queries taking a device in both interfaces. I am not sure if either backend will ever return different minimums, I suspect the actual requirement for the alignment and size comes into play when you map them onto physical memory, which are allocated on specific devices. Depending on how we should read the Level Zero requirement here, we could rephrase the reservation interface requirement to be that it must be aligned in accordance with the granularity of any device it will be mapped to. Arguably, this is more of an implicit requirement from the map function though.

Reply:
Since neither the Level Zero nor the CUDA API takes a device handle, I assume that both APIs must be reserving the address range in all devices contained by the context.  Would you agree?

In that case, wouldn't it make sense to remove the `device` parameter from `get_minimum_mem_granularity` and `get_recommended_mem_granularity` also?

Reply:
I would be okay with it, but what would it do if the devices report different granularity? I assume the best solution would be to try and find the smallest value that is a multiple of all the reported granularities.

Reply:
Let's ask @jandres742 about the Level Zero API.

The documentation for [zeVirtualMemReserve](https://spec.oneapi.io/level-zero/latest/core/api.html#zevirtualmemreserve) says:

> The starting address and size must be page aligned. See [zeVirtualMemQueryPageSize](https://spec.oneapi.io/level-zero/latest/core/api.html#ze__api_8h_1afe411cceb631a3a176463ee05fda3dd7)

However, `zeVirtualMemQueryPageSize` takes an `hDevice` parameter while `zeVirtualMemReserve` does not.  What does the statement I  quote above mean exactly?  Does it mean that the application must call `zeVirtualMemQueryPageSize` for every device in the `hContext` and find an address that is aligned properly for every one of those devices?

Reply:
@gmlueck . zeVirtualMemReserve doesn't take a device handle since the address range is not associated with any device at that moment, as @steffenlarsen , the device is important during the mapping.

So on zeVirtualMemReserve, an address must be aligned for the potential devices the application would like to map to eventually. If they are several, then yes, application would have to traverse them and find a starting address that suits the alignment of all of them.

Now, this is similar to CUDA behavior, where reserve doesnt take a device parameter, but getGranularity does. So removing the device handle from getGranularity and do what @steffenlarsen  says makes sense:

> I would be okay with it, but what would it do if the devices report different granularity? I assume the best solution would be to try and find the smallest value that is a multiple of all the reported granularities.

however, problem would be if we have an application in CUDA already querying granularities on all devices, when porting to SYCL, we would have have double traversal of devices and queries.

So in that sense, I would suggest to keep the device parameter for the getGranularity API.

Reply:
@jandres742: You seem to be saying that `zeVirtualMemReserve` does not have any alignment requirements and the only alignment requirement comes from `zeVirtualMemMap`.  If that is the case, is this statement in the Level Zero specification for `zeVirtualMemReserve` wrong:

> The starting address and size must be page aligned. See [zeVirtualMemQueryPageSize](https://spec.oneapi.io/level-zero/latest/core/api.html#ze__api_8h_1afe411cceb631a3a176463ee05fda3dd7).

Should that statement be removed from the spec?

Reply:
@gmlueck : sorry if my statement was confusing. zeVirtualMemReserve has alignment requirements, just as cuMemAddressReserve. What I'm saying is that zeVirtualMemReserve  doesnt take a device since the range may be mapped later to any device. So important thing is that the alignment used for zeVirtualMemReserve is good for any potential device the application would like to use, which again, I think it is similar to CU functionality.

Reply:
What exactly are the alignment requirements of `zeVirtualMemReserve`?  Must the address be correctly aligned for *every* device in the `hContext`?

Reply:
The issue in this thread still needs to be resolved.  This statement in the SYCL extension specification is unclear:

> The `uintptr_t` must be aligned in accordance with either the minimum or recommended granularity, as queried through `get_mem_granularity`

The call to `reserve_virtual_mem` takes a `syclContext` parameter, which may contain many devices whereas the call to `get_mem_granularity` takes a single `syclDevice`.  Is it the application's responsibility to call `get_mem_granularity` for each device in `syclContext` and then find an alignment that is in common with all devices in the context?  That seems like a big burden to place on the application.

I think the same problem exists in the L0 specification for `zeVirtualMemReserve`, which has this unclear statement:

> The starting address and size must be page aligned. See [zeVirtualMemQueryPageSize](https://spec.oneapi.io/level-zero/latest/core/api.html#ze__api_8h_1afe411cceb631a3a176463ee05fda3dd7)

Again, is it the caller's responsibility to call `zeVirtualMemQueryPageSize` for each device in `hContext` and find a page size that is in common for all devices?

Reply:
From the previous discussion, it sounds like the requirements only need to be enforced for any devices the user intends to map the virtual memory onto, so maybe the most user-friendly approach here would be to add a few more overloads to `get_mem_granularity` that can find a common granularity for a set of devices. Maybe something like:

```c++
size_t get_mem_granularity(const context &syclContext, const std::vector<device> &syclDevices,
                           granularity_mode mode = granularity_mode::recommended);
size_t get_mem_granularity(const context &syclContext,
                           granularity_mode mode = granularity_mode::recommended);
size_t get_mem_granularity(const queue &syclQueue,
                           granularity_mode mode = granularity_mode::recommended);
size_t get_mem_granularity(const physical_mem &syclPhysicalMem,
                           granularity_mode mode = granularity_mode::recommended);
```

where the one taking a context would just call the one before with the result of `syclContext.get_devices()`. With these, the user can still get the granularity for a specific device, either by just passing a single device to the first one or through the `queue`/`physical_mem` overloads (or if `syclContext` only has a single device, the second query as well.)

With those, we can specify the requirement for `reserve_virtual_mem` that the user must make sure to be using the appropriate granularity for all the devices in the corresponding context that they intend to map the reserved address range onto.

Reply:
> From the previous discussion, it sounds like the requirements only need to be enforced for any devices the user intends to map the virtual memory onto

The description for `reserve_virtual_mem` says that "the `uintptr_t` **must** be aligned ...".  I assume this means that the function will fail by throwing an exception if it is not properly aligned.  Since `reserve_virtual_mem` doesn't know which devices you will use, I assume it will diagnose an error if the address is misaligned for any of the devices.  If this is not the intent, the spec wording should be changed.

The same is true for the Level Zero specification of `zePhysicalMemCreate`.  The description says that the address **must** be aligned, so I would assume that the call fails if the address is not properly aligned.

Whenever we use the word "must" in the specification of an API, this places a requirement on the legal values the application may pass to that API.

> where the one taking a context would just call the one before with the result of `syclContext.get_devices()`.

It's more complicated than that.  I think it needs to find the least common multiple among all of the alignments reported by each device in the context.

Reply:
> The description for reserve_virtual_mem says that "the uintptr_t must be aligned ...". I assume this means that the function will fail by throwing an exception if it is not properly aligned. Since reserve_virtual_mem doesn't know which devices you will use, I assume it will diagnose an error if the address is misaligned for any of the devices. If this is not the intent, the spec wording should be changed.

That's not currently the case and I should probably change the wording here. I intentionally did not try to enforce it to avoid the overhead of having to query for the granularity each time the user reserves and then verify their alignment/sizes. Depending on what we decide here, I will go adjust accordingly.

> The same is true for the Level Zero specification of zePhysicalMemCreate. The description says that the address must be aligned, so I would assume that the call fails if the address is not properly aligned.

I don't know. I am not sure we can even have a case with our current devices, or at least I don't have a setup with multiple L0 devices that report different granularities. My understanding of @jandres742's previous comments are that the requirement isn't actually really applicable until we know which devices the virtual address will get mapped to, which we cannot know at the time of the reservation, except of course it being 0 or more of the devices in the context. Maybe I misunderstood or maybe the we want to ask for the L0 documentation for these requirements to be a made more explicit too.

> It's more complicated than that. I think it needs to find the least common multiple among all of the alignments reported by each device in the context.

Yeah sorry, that is what I meant. The first overload would find the least common multiple of the granularities and the second overload would call the first overload with all devices in the context, finding their least common multiple of their granularities.

Reply:
It sounds like you are saying that there is no requirement about the alignment of the `uintptr_t` that is passed to `reserve_virtual_mem`, and instead there is an alignment requirement for the `uintptr_t` that is passed to `physical_mem::map`.  If that is the case, the wording about the required alignment should be moved.

> maybe the we want to ask for the L0 documentation for these requirements to be a made more explicit too.

Yes, I think we need this.

Obviously, we cannot remove the alignment requirement from `reserve_virtual_mem` unless the underlying call to `zeVirtualMemMap` also has no alignment requirement.

Reply:
> What exactly are the alignment requirements of zeVirtualMemReserve? Must the address be correctly aligned for every device in the hContext?

@gmlueck . Not necessarily, because when you do memReserve you dont know to which you will map. But since at the memReserve time you dont know which devices would be used, so yes, having the address aligned to each device in the context is a good approach.

Reply:
> maybe the we want to ask for the L0 documentation for these requirements to be a made more explicit too.
> Yes, I think we need this.

@gmlueck : what is the documentation we have on this on CU?

Reply:
> what is the documentation we have on this on CU?

@steffenlarsen has been looking at the CU documentation for this, so he probably knows better.

> Not necessarily, because when you do memReserve you dont know to which you will map. But since at the memReserve time you dont know which devices would be used, so yes, having the address aligned to each device in the context is a good approach.

I'm not really asking about what a "good" approach is for applications to follow.  I'm asking what the hard requirement is from the L0 specification.  Does `zeVirtualMemReserve` have any hard requirements on the alignment of the `pStart` and `size` parameters?  Or, is it only `zeVirtualMemMap` that has an alignment requirement?

Reply:
The [CUDA documentation](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__VA.html#group__CUDA__VA_1ge489256c107df2a07ddf96d80c86cd9b) specifies that the requirement for reservations are:

> The size and address parameters must be a multiple of the host page size and the alignment must be a power of two or zero for default alignment.

I suppose it is outside the scope of it to specify how the user gets the host page size. It does not make any mention of the granularity at this level.

Reply:
@gmlueck :

> Does zeVirtualMemReserve have any hard requirements on the alignment of the pStart and size parameters?
> Or, is it only zeVirtualMemMap that has an alignment requirement?

Both, as as mentioned in the spec:

"The starting address and size must be page aligned. See [zeVirtualMemQueryPageSize](https://spec.oneapi.io/level-zero/latest/core/api.html?highlight=zevirtualmemreserve#ze__api_8h_1afe411cceb631a3a176463ee05fda3dd7)."

"The virtual start address and size must be page aligned. See [zeVirtualMemQueryPageSize](https://spec.oneapi.io/level-zero/latest/core/api.html?highlight=zevirtualmemreserve#ze__api_8h_1afe411cceb631a3a176463ee05fda3dd7)."

Reply:
This conversation never got resolved.  I think the core problem is that the following statement is unclear:

> `start` must be aligned in accordance with either the minimum or recommended granularity, as returned by a call to `get_mem_granularity`. Likewise, `numBytes` must be a multiple of the granularity.

The function `get_mem_granularity` returns the allocation granularity *of a particular device*, however `reserve_virtual_mem` does not take a device parameter.  Therefore, it's not clear what granularity we are talking about in the statement I quote above.

I think the solution might be to remove the `syclDevice` parameter from `get_mem_granularity`.  As a result, `get_mem_granularity` would return the allocation granularity for a context (not a device).  This solves the problem because `reserve_virtual_mem` also takes a context.

However, this probably requires a change to Level Zero because `zeVirtualMemQueryPageSize` returns a page size for a particular device, not for a context.  See my comments in this thread above, though.  I think the Level Zero API also has problems, which would be solved by changing the definition of `zeVirtualMemQueryPageSize`.

Reply:
Examining Level Zero code:  the zeVirtualMemQueryPageSize() API calls the following internal function:

ze_result_t ContextImp::queryVirtualMemPageSize(ze_device_handle_t hDevice,
                                                size_t size,
                                                size_t *pagesize) {
    // Retrieve the page size and heap required for this allocation size requested.
    getPageAlignedSizeRequired(size, nullptr, pagesize);
    return ZE_RESULT_SUCCESS;
}

So, the hDevice handle is not used.   Going down the call tree, the following is eventually called:

size_t DrmMemoryManager::selectAlignmentAndHeap(size_t size, HeapIndex *heap) {
    AlignmentSelector::CandidateAlignment alignmentBase = alignmentSelector.selectAlignment(size);
    size_t pageSizeAlignment = alignmentBase.alignment;
    auto rootDeviceCount = this->executionEnvironment.rootDeviceEnvironments.size();

    // If all devices can support HEAP EXTENDED, then that heap is used, otherwise the HEAP based on the size is used.
    for (auto rootDeviceIndex = 0u; rootDeviceIndex < rootDeviceCount; rootDeviceIndex++) {
        auto gfxPartition = getGfxPartition(rootDeviceIndex);
        if (gfxPartition->getHeapLimit(HeapIndex::heapExtended) > 0) {
            auto alignSize = size >= 8 * MemoryConstants::gigaByte && Math::isPow2(size);
            if (debugManager.flags.UseHighAlignmentForHeapExtended.get() != -1) {
                alignSize = !!debugManager.flags.UseHighAlignmentForHeapExtended.get();
            }

            if (alignSize) {
                pageSizeAlignment = Math::prevPowerOfTwo(size);
            }

            *heap = HeapIndex::heapExtended;
        } else {
            pageSizeAlignment = alignmentBase.alignment;
            *heap = alignmentBase.heap;
            break;
        }
    }
    return pageSizeAlignment;

So all detected devices are cycled in this loop.   I verified this to be the case using a board with 4x ATS-M devices and verified that all 4 devices were looped.   Based upon this, only one call to zeVirtualMemQueryPageSize() should be required for the given driver.

Reply:
Any chance we can change the parameters of `zeVirtualMemQueryPageSize` to remove `hDevice` or document that the `hDevice` parameter isn't used?  It would be nice if SYCL could rely on documented behavior of Level Zero, rather than making an assumption based on the current implementation.